pkg/reconciler/namespace/namespace.go

```diff
@@ -0,0 +1,191 @@
+
+	original, err := c.nsLister.Get(key)
+
+	if apierrs.IsNotFound(err) {
+		logger.Errorf("Namespace %s in work queue no longer exists %s", key, err)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy
+	newNamespace := original.DeepCopy()
+
+	whitelist, exists := newNamespace.Labels[whitelistLabel]
+	if !exists || whitelist == "false" {
+		logger.Infof("Skipping namespace certificate for namespace %s", newNamespace.Name)
+		return nil
+	}
+
+	err = c.reconcile(ctx, newNamespace)
+	if err != nil {
+		c.Recorder.Event(newNamespace, corev1.EventTypeWarning, "InternalError", err.Error())
+	}
+	return err
+}
+
+func (c *Reconciler) reconcile(ctx context.Context, ns *corev1.Namespace) error {
+	cfg := config.FromContext(ctx)
+
+	// Only create wildcard certs for the default domain
+	defaultDomain := cfg.Domain.LookupDomainForLabels(nil)
+	if defaultDomain == "" {
+		c.Logger.Infof("Skipping wildcard certificate creation for namespace %s. No default domain defined.", ns.Name)
+		return nil
+	}
+
+	domain, err := getWildcardDomain(cfg.Network.DomainTemplate, defaultDomain, ns.Name)
+	if err != nil {
+		c.Logger.Errorf("Failed to apply domain template %s to domain %s and namespace %s", cfg.Network.DomainTemplate, defaultDomain, ns.Name)
+	}
+
+	desiredCert := makeDesiredCertificate(ns, []string{domain})
+
+	certs, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(ns.Name).List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	existingCert, err := getExistingCert(domain, certs)
+
+	if apierrs.IsNotFound(err) {
+		cert, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(ns.Name).Create(desiredCert)
+		if err != nil {
+			c.Logger.Errorw("Failed to create namespace certificate", zap.Error(err))
+			c.Recorder.Eventf(ns, corev1.EventTypeWarning, "Creation Failed",
+				"Failed to create Knative certificate %s/%s: %v", ns.Name, desiredCert.ObjectMeta.Name, err)
+			return nil
+		}
+		c.Recorder.Eventf(cert, corev1.EventTypeNormal, "Created",
+			"Created Knative Certificate %s/%s", ns.Name, cert.ObjectMeta.Name)
+	} else if err != nil {
+		return err
+	} else if !equality.Semantic.DeepEqual(existingCert.Spec, desiredCert.Spec) {
+		copy := existingCert.DeepCopy()
+		copy.Spec = desiredCert.Spec
+		_, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(copy.Namespace).Update(copy)
+		if err != nil {
+			c.Logger.Errorw("Failed to update Knative Certificate", zap.Error(err))
+			c.Recorder.Eventf(existingCert, corev1.EventTypeWarning, "UpdateFailed",
+				"Failed to update Knative Certificate %s/%s: %v", desiredCert.Namespace, desiredCert.Name, err)
+			return err
+		}
+		c.Recorder.Eventf(existingCert, corev1.EventTypeNormal, "Updated",
+			"Updated Spec for Knative Certificate %s/%s", desiredCert.Namespace, desiredCert.Name)
+		return nil
+	}
+
+	return nil
+}
+
+func makeDesiredCertificate(namespace *corev1.Namespace, dnsNames []string) *v1alpha1.Certificate {
```

Comment:
It seems we always call the function with a single parameter, so we probably should just take in `string` and make the slice in the actual `Certificate` object construction.